Serato: Import Beatgrid from metadata

Holzhaus opened the pull request:
I finally found a few hours to implement parsing of the Serato Beatgrid. This already works, ~~but nothing is imported yet, because we have the same problem as we had with cues, i.e. we can't determine the timing offset at the time when the metadata is read. We need to find a nice workaround for that.~~

- [x] Add more tests
- [x] Actually import the beatgrid while taking timing offsets into account

Support for tag types:
  - [X] ID3v2
  - [x] FLAC/Xiph
  - [ ] ~~Ogg/Xiph~~ *Format differs significantly, no idea how to parse this.*
  - [x] MP4

**EDIT:** It works! here are two tracks with their beatgrids imported from Serato (for demo purposes I added some tempo changes to the upper track's beatgrid)
![2020-07-22-172204_1231x378_scrot](https://user-images.githubusercontent.com/1834516/88195571-7b7d3d80-cc40-11ea-8da6-9b8db9619169.png)

Holzhaus committed:
track/serato: Add parser for "Serato BeatGrid" tag

Holzhaus committed:
track/taglib: Read "Serato BeatGrid" ID3 tag from track metadata

Holzhaus committed:
track/serato/beatgrid: Store terminal markers separately

Holzhaus committed:
track/serato/beatgrid: Add method to get list of actual beat positions

Holzhaus committed:
track/serato/beatgrid: Add comment regarding dumpID3 data size

Holzhaus committed:
track/serato/beatgrid: Add more checks for data validity

Holzhaus committed:
track/serato/beatgrid: Fix typo in member variable

Be-ing commented:
I think this should go to 2.3, not master.

Holzhaus commented:
> I think this should go to 2.3, not master.

This is based on 2.3, so you can simply change the target branch. I wan't sure about it, because we're already in feature freeze.

Be-ing commented:
I consider this finishing an incomplete feature.

Holzhaus commented:
Btw, I added the current state of Serato Mrtadata support in Mixxx here: https://github.com/mixxxdj/mixxx/wiki/Serato-Metadata-Format#mixxx-support-status

Holzhaus committed:
test: Add some simple roundtrip tests for Serato Beatgrid parsing

Holzhaus committed:
track: Add SeratoBeatsImporter for deferred beatmap import

Holzhaus committed:
track: Implement deferred import of beats (like we do for cues)

Holzhaus committed:
track/serato/beatgrid: Fix usage of wrong unit in Serato BeatGrid code

Holzhaus committed:
track/track: Fix unused variable warning (only used in DEBUG_ASSERT)

Holzhaus committed:
track/taglib/trackmetadata_id3v2: Add support for Serato BeatGrid export

Holzhaus committed:
track/serato/beatgrid: Add support for FLAC tags

Holzhaus committed:
track/serato/beatgrid: Add support for for MP4 tags

Holzhaus committed:
test: Improve code readability in Serato BeatGrid test

Holzhaus commented:
Any comments?

daschuer commented on src/track/serato/beatgrid.cpp:
```diff
@@ -0,0 +1,401 @@
+                              << previousBeatPositionMillis;
+            return false;
+        }
+        previousBeatPositionMillis = pNonTerminalMarker->positionSecs();
```
fix naming: Millis vs Secs

daschuer commented on src/track/serato/beatgrid.cpp:
```diff
@@ -0,0 +1,401 @@
+    // For now, we only read it to be able to dump the exact same byte sequence
+    // later on.
+    quint8 footer;
+    stream >> footer;
```
I can imagine Serato uses it as "dirty" flag. I am afraid we need to toggle it as well in case we alter the beat grid in Mixxx. Can you verify it?

Holzhaus replied:
I can check if it new values are only picked up by Serato if that value is different (if that's what you mean). However, Serato is also doing strange stuff elsewhere (e.g. they sometimes append a random byte for base64 encoded flac metadata). My guess was that they have a messed up length check and write a  byte of uninitialized memory :D

daschuer replied:
That seems also be likely, in particularly if it is not constantly increasing.

Holzhaus replied:
It's not increasing.

daschuer commented on src/track/serato/beatgrid.cpp:
```diff
@@ -0,0 +1,401 @@
+    return true;
+}
+
+QByteArray SeratoBeatGrid::dump(taglib::FileType fileType) const {
```
dump to what? Does it make sense to immediately create the TagLib::String()? 
Than the function can be named dumpToTagLibString() 

Note: TagLib::String() uses internally a TagLib::wstring just like QString. 

Not sure if we can make use of it, by configuring the QDataStream to generate wchars directly ... mm

All these considerations are probably not worth the time :-/

Holzhaus replied:
> dump to what? Does it make sense to immediately create the TagLib::String()?

No, I think makes sense to dump it to a plain QByteArray. Depending on the tag type this can be written to a `TagLib::String` or a `TagLib::ByteVector`. Also, it makes testing easier and if if we ever switch to another tag library we do not need to touch this. All taglib-related code remains in `src/track/taglib`.

daschuer replied:
OK. 
I was originally really struggling to what the value is dumped.  
Can we find a better function name or add a comment?

Holzhaus replied:
I'll add a comment.

daschuer commented on src/track/taglib/trackmetadata_common.cpp:
```diff
@@ -188,6 +211,16 @@ bool parseSeratoMarkers2(
+        FileType fileType) {
+    const QByteArray utf8Data =
+            trackMetadata.getTrackInfo().getSeratoTags().dumpBeatGrid(fileType);
+    return TagLib::String(
```
This can make use of a comment about the format of data we handle here.

daschuer replied:
Do we have here the base64 format? I think it would also helpful to append the dump() function with a to... prefix to make it more maintainable.

Holzhaus replied:
Can you elaborate what you mean? I'm not sure. The Base64 encoding is only used if `fileType` is FLAC or MP4, not MP3.

daschuer replied:
utf8Data can be everything. I don't even think it is actually utf8. So this is equivalent to "var1" or such. 
A bit info about the content and content format is helpful here.

daschuer commented on src/track/track.cpp:
```diff
@@ -1374,18 +1484,48 @@ void Track::updateAudioPropertiesFromStream(
+
+    if (importBeats) {
+        kLogger.debug()
+                << "Finishing deferred import of beats";
```
Why was it deferred and where?

Holzhaus replied:
Because the positions depend on the tracks sample rate. Here:
https://github.com/mixxxdj/mixxx/pull/2958/files/2eba4e80bc28361a826d30654858862b591e9bb6#diff-f59a0f669757d180c782c7a8fa0adea0R933-R946

daschuer replied:
OK, so the comment can become: 
"Finishing import because we know now track sample rate and offset"

daschuer reviewed (changes requested):
Thank you for this PR, I have left some comments.

Holzhaus commented on src/track/beatsimporter.h:
```diff
@@ -0,0 +1,34 @@
+
+    /// Determines the timing offset and returns a Vector of frame positions
+    /// to use as input for the BeatMap constructor
+    virtual QVector<double> importBeatsWithCorrectTiming(
```
> Which timing is correct depends on the use case.

What do you mean? The timing of *imported* beats/cues is correct if and only if the beats are at the exact same positions as in Serato. Or what do you mean?

> The function also dos not the import, it reads and returns a adjusted beats vector.
>
> So how about getBeatsWithMixxxTiming() or such.

I think the naming for beats and cue import should be consistent, and the "import" name was requested by @uklotzde here: https://github.com/mixxxdj/mixxx/pull/2707#discussion_r415381229

daschuer replied:
I think I disagree here with @uklotzde, because import implies also storing. This function does not store anything. 
But there is no strong need to change it. 

> The timing of imported beats/cues is correct if and only if the beats are at the exact same positions as in Serato

Or 

> The timing of imported beats/cues is correct if they match the values in Serato. 

We have no definition of "CorrectTiming"

Holzhaus replied:
> The timing of imported beats/cues is correct if they match the values in Serato.

They already match the values in Serato without this function. This function corrects the offset that is caused by Mixxx using a different decoder. Unfortunately there is no "Mixxx" timing as is depends on the decoder that is currently used.

daschuer replied:
OK, so it might become:

importBeatsWithOffsetForCurrentEncoder() 
 
?

Holzhaus replied:
I renamed it to "importBeatsAndApplyTimingOffset".

Holzhaus commented on src/track/serato/beatgrid.cpp:
```diff
@@ -0,0 +1,401 @@
+QByteArray base64encode(const QByteArray& data, bool chopPadding) {
+    QByteArray dataBase64;
+
+    // A newline char is inserted at every 72 bytes of base64-encoded content.
```
By this function, which mirrors what Serato does.

daschuer replied:
Ok, so please update the comment.

Holzhaus commented on src/track/serato/beatgrid.cpp:
```diff
@@ -0,0 +1,401 @@
+            return false;
+        }
+
+        QByteArray markerData = QByteArray(buffer, kMarkerSizeID3);
```
It's not really a buffer, I'll just replace it with QByteArray::fromRawData to avoid a deep copy.

Holzhaus committed:
track/serato/beatgrid: Improve comments regard base64 encoding

Holzhaus committed:
track/serato/beatgrid: Use QByteArray::reserve() instead of resize()

Holzhaus committed:
track/serato/beatgrid: Fix wrong unit in variable name

Holzhaus committed:
track/serato/beatgrid: Add missing "// static" comments

Holzhaus committed:
track/serato/beatgrid: Use fromRawData to avoid deep copies when parsing

Holzhaus commented on src/track/serato/beatgrid.cpp:
```diff
@@ -0,0 +1,401 @@
+
+QByteArray SeratoBeatGridNonTerminalMarker::dumpID3() const {
+    QByteArray data;
+    data.resize(kMarkerSizeID3);
```
I replaced it with reserve.

Holzhaus committed:
track/taglib/trackmetadata_commont: Improve var names for Serato data

Holzhaus committed:
track/serato/beatgrid: Add some comments and make some method private

Holzhaus committed:
track/serato/beatgrid: Improve Base64 prefix parsing

Holzhaus committed:
sources/soundsourceproxy: Improve debug assertion for Serato import

Holzhaus committed:
Merge branch '2.3' of github.com:mixxxdj/mixxx into serato-beatgrid-import

Holzhaus committed:
track/serato/beatgrid: Use doxygen-style /// comments in header

Holzhaus commented on src/track/track.cpp:
```diff
@@ -900,26 +916,104 @@ void Track::setCuePoints(const QList<CuePointer>& cuePoints) {
+        importPendingBeatsMarkDirtyAndUnlock(&lock);
+        return ImportStatus::Complete;
+    } else {
+        if (m_pBeatsImporterPending->isEmpty()) {
+            // Just return the current import status without clearing any
+            // existing cue points.
+            return (!m_pBeatsImporterPending->isEmpty())
+                    ? ImportStatus::Complete
+                    : ImportStatus::Pending;
+        }
```
This is basically copy/pasted from https://github.com/mixxxdj/mixxx/pull/2958/files#diff-f59a0f669757d180c782c7a8fa0adea0R1011-R1017.

@uklotzde Shouldn't this always return `Complete`? Right now I think it always returns `Pending`, right?

uklotzde replied:
Oh, this check redundant. Yes, please fix it and return Complete when empty. Might be a remainder from a previous refactoring.

Holzhaus committed:
test/seratobeatgridtest: Improve roundtrip test and write output on fail

Holzhaus committed:
track/serato/beatgrid: Implement support for lossless MP4/FLAC roundtrip

Holzhaus committed:
test: Add roundtrip tests for Serato Beatgrid data in MP4/FLAC files

Holzhaus commented on src/track/serato/beatgrid.cpp:
```diff
@@ -0,0 +1,401 @@
+            // In case that the last block would require padding, Serato seems to
+            // chop off the last byte of the base64-encoded data
+            if (block.size() % 3) {
+                dataBase64.chop(1);
```
Your argument is based on logic and the rules of valid base64 encoding, which not what Serato's data format is about :wink:

Do you have a sample file from Serato where this is the case? Otherwise I'd say we leave it as is. I did some extensive testing when I originally I originally implemented this, so I'd rather leave this as is.

daschuer replied:
No, I have no Serato installation.

Ok, than please add your test results as a comment. 
Interesting is: 
Did, you have a test file where Serato adds a zero byte where two bytes would be sufficient?  
Does Serato except a zero byte when two bytes are sufficient?

Holzhaus committed:
track/track: Improve deferred import debug messages

Holzhaus committed:
track/beatsimporter: Rename method to importBeatsAndApplyTimingOffset

Holzhaus committed:
track/cueinfoimporter: Rename method to importCueInfosAndApplyTimingOffset

Holzhaus commented:
@daschuer Thanks for reviewing. As you're the only reviewer, you can hit merge :D

daschuer commented:
OK, I have not hit merge because not all Todos are checked.

daschuer commented:
CI is also not finished.

uklotzde commented on src/track/beatsimporter.h:
```diff
@@ -0,0 +1,34 @@
+    BeatsImporter() = default;
+    virtual ~BeatsImporter() = default;
+
+    virtual bool isEmpty() const {
```
Why not pure virtual to prevent forgetting to override this method in derived classes?

uklotzde commented on src/track/beatsimporter.h:
```diff
@@ -0,0 +1,34 @@
+
+    /// Determines the timing offset and returns a Vector of frame positions
+    /// to use as input for the BeatMap constructor
+    virtual QVector<double> importBeatsAndApplyTimingOffset(
```
Same here, empty default implementations are dangerous even if they are able to return some result.

uklotzde commented on src/track/serato/beatgrid.h:
```diff
@@ -0,0 +1,185 @@
+            : m_positionSecs(positionSecs),
+              m_beatsTillNextMarker(beatsTillNextMarker) {
+    }
+    ~SeratoBeatGridNonTerminalMarker() = default;
```
Unneeded

uklotzde commented on src/track/serato/beatgrid.h:
```diff
@@ -0,0 +1,185 @@
+            : m_positionSecs(positionSecs),
+              m_bpm(bpm) {
+    }
+    ~SeratoBeatGridTerminalMarker() = default;
```
Unneeded. Instead declarate the class as `final`.

uklotzde commented on src/track/serato/beatgrid.h:
```diff
@@ -0,0 +1,185 @@
+    float m_bpm;
+};
+
+inline QDebug operator<<(QDebug dbg, const SeratoBeatGridTerminalMarker& arg) {
```
I prefer to put these non-performance-critical utility functions into the .cpp file instead of cluttering the header file. Also when inlined the string literals might be duplicated on every call site (didn't check if the compiler is smart enough to recognize the redundancy).

uklotzde commented on src/track/serato/beatgrid.h:
```diff
@@ -0,0 +1,185 @@
+class SeratoBeatGrid final {
+  public:
+    SeratoBeatGrid() = default;
+    explicit SeratoBeatGrid(
```
Doesn't need to be explicit. Inconsistent C++ weirdness. Doesn't matter but I always stumble upon this when scanning the number of arguments  ;)

uklotzde commented on src/track/serato/beatgrid.h:
```diff
@@ -0,0 +1,185 @@
+        m_nonTerminalMarkers = nonTerminalMarkers;
+    }
+
+    const SeratoBeatGridTerminalMarkerPointer terminalMarker() const {
```
Redundant const on return type

uklotzde commented on src/track/track.cpp:
```diff
@@ -900,26 +916,104 @@ void Track::setCuePoints(const QList<CuePointer>& cuePoints) {
             return (!m_pCueInfoImporterPending->isEmpty())
-                    ? CueImportStatus::Complete
-                    : CueImportStatus::Pending;
+                    ? ImportStatus::Complete
```
Same here: Remove redundant check and return Complete.

uklotzde reviewed (changes requested):
Sorry, for the delay. Just some minor issues.

Holzhaus committed:
track/beatsimporter: Remove default implementation for virtual methods

Holzhaus committed:
track/serato/beatgrid: Make class final and remove default destructor

Holzhaus committed:
track/serato/beatgrid: Remove uneeded explicit from constructor

Holzhaus committed:
track/serato/beatgrid: Remove unneeded const from return type

Holzhaus committed:
track/serato/beatgrid: Move QDebug methods to cpp file

Holzhaus committed:
track/track: Remove redundant check if importer is empty

Also, in case the importer is empty this should return "complete", not
"pending".

Holzhaus committed:
track/serato/beatsimporter: Improve isEmpty check

A terminal marker always need to be present to generate a beatgrid.

uklotzde commented on src/track/beatsimporter.h:
```diff
@@ -0,0 +1,27 @@
+/// signal info (channel number, sample rate, bitrate) is known.
+class BeatsImporter {
+  public:
+    BeatsImporter() = default;
```
Redundant default ctor

uklotzde commented on src/track/track.cpp:
```diff
@@ -900,26 +916,100 @@ void Track::setCuePoints(const QList<CuePointer>& cuePoints) {
+        // points immediately
+        importPendingBeatsMarkDirtyAndUnlock(&lock);
+        return ImportStatus::Complete;
+    } else {
```
Transforming this into `else if` + `else` could improve the readability with all 3 cases aligned.

uklotzde commented on src/track/track.cpp:
```diff
@@ -900,26 +916,100 @@ void Track::setCuePoints(const QList<CuePointer>& cuePoints) {
+            // existing cue points.
+            return ImportStatus::Complete;
+        }
+        DEBUG_ASSERT(!m_pBeatsImporterPending->isEmpty());
```
redundant debug assert

uklotzde commented on src/track/track.cpp:
```diff
@@ -900,26 +916,100 @@ void Track::setCuePoints(const QList<CuePointer>& cuePoints) {
         importPendingCueInfosMarkDirtyAndUnlock(&lock);
-        return CueImportStatus::Complete;
+        return ImportStatus::Complete;
     } else {
```
Also here: Use `else if` + `else` and remove redundant debug assert

Holzhaus committed:
track/beatsimporter: Remove redundant default ctor

Holzhaus committed:
track/track: Improve readability of importBeats/CueInfos methods

uklotzde commented:
Verified locally. LGTM
